File: the-daily-starter/pages/api/user/delete.ts
import { NextApiRequest, NextApiResponse } from "next";
import { deleteUserProfileAndUsername } from "@/lib/firebase/adminUtils/firebaseUserService";
import { admin } from "@/lib/firebase/firebaseAdmin";

export default async function handler(req: NextApiRequest, res: NextApiResponse){

    if(req.method !== 'DELETE'){
        res.status(405).json({ message: "Method Not Allowed." })
    }

    try{
        const authHeader = req.headers.authorization;
        if(!authHeader || typeof authHeader !== "string" || !authHeader.startsWith("Bearer ")){
            return res.status(401).json({ message: "Unauthorized." });
        };

        const idToken = authHeader.split("Bearer ")[1];
        const decodedToken = await admin.auth().verifyIdToken(idToken);

        const requesterUid = decodedToken.uid;
        const isAdmin = decodedToken.admin === true;

        const { uid } = req.body as { uid?: string };

        if (!uid) {
            return res.status(400).json({ error: "Missing uid" });
        }

        //Check permissions 
        if (!isAdmin && requesterUid !== uid) {
            return res.status(403).json({ error: "Permission denied" });
        }

        await deleteUserProfileAndUsername(uid);

        return res.status(200).json({ message: `User ${uid} deleted successfully` });

    } catch (err: any) {
        console.error("An error occurred while attempting to delete this account: " + err);
        return res.status(500).json({ error: err.message || "Error while deleting account" });
    }
}

File: backend/supabase/userservice/endpoint/delete.go
package SupabaseUserEndpoint

import (
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	SupabaseUsers "services/supabase/userservice"
)

type DeleteUserRequest struct {
	UID string `json:"uid"`
}

func DeleteUserHandler(svc *SupabaseUsers.SupabaseService) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodDelete {
			http.Error(w, "Method Not Allowed", http.StatusMethodNotAllowed)
		}

		uid, ok := r.Context().Value("uid").(string)
		if !ok || uid == "" {
			http.Error(w, "Unauthorized.", http.StatusUnauthorized)
		}

		isAdmin, _ := r.Context().Value("isAdmin").(bool)

		var reqBody DeleteUserRequest
		if err := json.NewDecoder(r.Body).Decode(&reqBody); err != nil {
			log.Printf("Error decoding request body: %v", err)
			http.Error(w, "Bad Request.", http.StatusBadRequest)
			return
		}

		if reqBody.UID == "" {
			http.Error(w, "Missing uid", http.StatusBadRequest)
		}

		if !isAdmin && uid != reqBody.UID {
			http.Error(w, "Permision denied.", http.StatusForbidden)
			return
		}

		if err := svc.DeleteUserProfileSupabase(r.Context(), reqBody.UID); err != nil {
			log.Printf("An error occurred while attempting to delete this account: %v", err)
			http.Error(w, "Error while deleting account.", http.StatusInternalServerError)
			return
		}

		w.WriteHeader(http.StatusOK)
		w.Header().Set("Content-Type", "application/json")
		fmt.Fprintf(w, `{"message": "User %s deleted successfully}`, uid)
	}
}
